### ISSUE TITLE
Incorrect retrieval results and inflated similarity scores for sdadas/stella-pl-retrieval-8k in TEI

### ISSUE THREAD

**Issue by Moraw1993** [huggingface/text-embeddings-inference#874]
## Description

We observed incorrect semantic retrieval behavior when serving `sdadas/stella-pl-retrieval-8k` with Text Embeddings Inference.

The same retrieval pipeline works correctly with `SentenceTransformers`, but produces degraded results when embeddings are generated by TEI.

The issue appears to be specific to `sdadas/stella-pl-retrieval-8k`, because `sdadas/stella-pl-retrieval-mini-8k` works correctly in the same setup.

## Models tested

Problematic model:

```text
sdadas/stella-pl-retrieval-8k
```

Control model that works correctly:

```text
sdadas/stella-pl-retrieval-mini-8k
```

Both models use custom implementation / custom code, so the issue does not seem to be caused by custom code in general.

## Environments

The issue was reproduced in two independent TEI deployments:

```text
Environment 1:
Runtime: Text Embeddings Inference
Hardware: NVIDIA Spark
Architecture: ARM64
Model: sdadas/stella-pl-retrieval-8k
Result: incorrect retrieval results
```
TEI /info
```json
{
  "model_id": "sdadas/stella-pl-retrieval-8k",
  "model_sha": null,
  "model_dtype": "float16",
  "served_model_name": "sdadas/stella-pl-retrieval-8k",
  "model_type": {
    "embedding": {
      "pooling": "mean"
    }
  },
  "max_concurrent_requests": 512,
  "max_input_length": 8192,
  "max_batch_tokens": 16384,
  "max_batch_requests": null,
  "max_client_batch_size": 32,
  "auto_truncate": true,
  "tokenization_workers": 19,
  "version": "1.9.3",
  "sha": "5bc4d889c38cf9c75e63617d62779bc0f6628b23",
  "docker_label": "sha-5bc4d88"
}
```

---

```text
Environment 2:
Runtime: Text Embeddings Inference
Hardware: NVIDIA GeForce RTX 4070
Architecture: x86_64
Model: sdadas/stella-pl-retrieval-8k
Result: same incorrect retrieval results
```

Reference implementation:

```text
Runtime: SentenceTransformers
Model: sdadas/stella-pl-retrieval-8k
Result: correct retrieval results
```

## Actual behavior

When using `sdadas/stella-pl-retrieval-8k` through TEI:

1. similarity scores are inflated across the corpus, often ranging from ~0.8 to ~0.9 even when searching over around 10k documents,
2. unrelated documents receive high similarity scores,
3. retrieval ranking is semantically incorrect,
4. the score distribution is not discriminative enough for semantic search.

Example:

A query related to programming jobs, software development or technical skills can return job offers such as excavator operator positions.

This is semantically incorrect because the query intent is related to programming, while the returned document is related to manual/operator work.

## Comparison with SentenceTransformers

For the same model and the same input texts:

```text
Model: sdadas/stella-pl-retrieval-8k
TEI vs SentenceTransformers embedding similarity: ~0.77
Expected similarity: ~0.99
```

With SentenceTransformers, retrieval ranking is correct and the score distribution looks reasonable.

With TEI, unrelated documents can receive high similarity scores and appear high in the ranking.

## Control test with mini model

We repeated the same test with:

```text
sdadas/stella-pl-retrieval-mini-8k
```

Results:

```text
TEI vs SentenceTransformers embedding similarity: ~0.99
Retrieval ranking: correct
Score distribution: correct
```

This suggests that the issue is not caused by TEI deployment in general, Spark/ARM64 specifically, or custom code in general. It seems specific to the full `sdadas/stella-pl-retrieval-8k` model served through TEI.

## Impact

This issue makes `sdadas/stella-pl-retrieval-8k` unreliable for production semantic search when served through TEI.

In our case, Polish job search returns semantically unrelated offers with high similarity scores, for example programming-related queries returning operator/manual job offers.

Please let us know what additional diagnostics would be useful. We can provide TEI startup command, Docker image version, score distributions, similarity matrices and example retrieved documents.

### Information

- [x] Docker
- [ ] The CLI directly

### Tasks

- [ ] An officially supported command
- [ ] My own modifications

### SentenceTransformers

```python
from sentence_transformers import SentenceTransformer
from sklearn.metrics.pairwise import cosine_similarity

model = SentenceTransformer(
    "sdadas/stella-pl-retrieval-8k",
    trust_remote_code=True,
)

texts = [
    "Szukam pracy jako programista. Mam doświadczenie w SQL, Pythonie, bazach danych i tworzeniu aplikacji.",
    "Programista aplikacji backendowych. Wymagana znajomość Python, SQL, REST API oraz baz danych.",
    "Operator koparki. Praca przy robotach ziemnych, obsługa koparki jednonaczyniowej oraz prace budowlane.",
    "Pracownik gospodarczy. Sprzątanie pomieszczeń, utrzymanie porządku, proste prace fizyczne.",
]

embeddings = model.encode(
    texts,
    normalize_embeddings=True,
)

print(cosine_similarity(embeddings))
```

### TEI

```bash
text-embeddings-router --model-id sdadas/stella-pl-retrieval-8k
```

Example request:

```bash
curl -X POST "http://localhost:8080/embed" \
  -H "Content-Type: application/json" \
  -d '{
    "inputs": [
      "Szukam pracy jako programista. Mam doświadczenie w SQL, Pythonie, bazach danych i tworzeniu aplikacji.",
      "Programista aplikacji backendowych. Wymagana znajomość Python, SQL, REST API oraz baz danych.",
      "Operator koparki. Praca przy robotach ziemnych, obsługa koparki jednonaczyniowej oraz prace budowlane.",
      "Pracownik gospodarczy. Sprzątanie pomieszczeń, utrzymanie porządku, proste prace fizyczne."
    ]
  }'
```

Then compare:

1. embeddings from TEI and SentenceTransformers for identical inputs,
2. cosine similarity matrix between texts,
3. retrieval ranking for the query about programming jobs.

## Expected behavior

TEI should generate embeddings close to SentenceTransformers for the same model, same input texts, same query/document formatting, same pooling strategy and same normalization settings.

For identical inputs, cosine similarity between TEI and SentenceTransformers embeddings should be close to `1.0`.

A programming-related query should rank programming-related job offers above unrelated manual/operator job offers.

**Comment by alvarobartt** [comment]
Thanks for reporting @Moraw1993, looking into it!
